spring-cloud-gcp-pubsub-stream-binder/src/test/java/org/springframework/cloud/gcp/stream/binder/pubsub/PubSubEmulator.java

```diff
@@ -0,0 +1,160 @@
+			Paths.get(".config", "gcloud", "emulators", "pubsub", "env.yaml"));
+
+	private static final Log LOGGER = LogFactory.getLog(PubSubEmulator.class);
+
+
+	// Binder to Pub/Sub emulator for use in individual tests.
+	private PubSubTestBinder binder;
+
+	// Reference to emulator instance, for cleanup.
+	private Process emulatorProcess;
+
+	// Hostname for cleanup, should always be localhost.
+	private String emulatorHost;
+
+	// Port for cleanup.
+	private String emulatorPort;
+
+	/**
+	 * Launches an instance of pubsub simulator, creates a binder.
+	 *
+	 * @throws Throwable for any failed setup.
+	 */
+	@Override
+	protected void before() throws Throwable {
+
+		this.emulatorProcess = new ProcessBuilder("gcloud", "beta", "emulators", "pubsub", "start").start();
+		assertConfigPresent();
+
+		Process envInitProcess = new ProcessBuilder("gcloud", "beta", "emulators", "pubsub", "env-init").start();
+
+		// env-init output is a shell command of the form "export PUBSUB_EMULATOR_HOST=localhost:8085".
+		String emulatorInitString
+				= new BufferedReader(new InputStreamReader(envInitProcess.getInputStream())).readLine();
+		envInitProcess.waitFor();
+		String emulatorHostPort = emulatorInitString.substring(emulatorInitString.indexOf('=') + 1);
+		extractTeardownParams(emulatorHostPort);
+
+		this.binder = new PubSubTestBinder(emulatorHostPort);
+	}
+
+	/**
+	 * Shuts down the two emulator processes.
+	 *
+	 * gcloud command is shut down through the direct process handle.
+	 * java process is identified and shut down through shell commands.
+	 *
+	 * There should normally be only one process with that host/port combination, but if there are more, they will be
+	 * cleaned up as well.
+	 */
+	@Override
+	protected void after() {
+		this.emulatorProcess.destroy();
+		try {
+			String hostPortParams = String.format("--host=%s --port=%s", this.emulatorHost, this.emulatorPort);
+			Process psProcess = new ProcessBuilder("ps", "-v").start();
+			new BufferedReader(new InputStreamReader(psProcess.getInputStream()))
+					.lines()
+					.filter(psLine -> psLine.contains(hostPortParams))
+					.map(psLine -> psLine.substring(0, psLine.indexOf(' ')))
+					.forEach(pid -> {
+						this.killProcess(pid);
+					});
+		}
+		catch (IOException e) {
+			LOGGER.error("Failed to cleanup: ", e);
+		}
+	}
+
+	/**
+	 * Returns an already-configured emulator binder when called from within a JUnit method.
+	 */
+	public PubSubTestBinder getBinder() {
+		return this.binder;
+	}
+
+	/**
+	 * Waits until a PubSub emulator configuration file is present.
+	 *
+	 * Fails if the file does not appear after 1 second.
+	 *
+	 * @throws InterruptedException which should interrupt the peaceful slumber and bubble up to fail the test.
+	 */
+	private void assertConfigPresent() throws InterruptedException {
+		int attempts = 10;
+		while (!Files.exists(this.emulatorConfigPath) && --attempts >= 0) {
```

Comment:
Is this basically verifying that the emulator started?
It seems somewhat fragile. For example, will it work on Windows?
Can we instead capture the stdout of the process to detect if it started successfully?

Reply:
This file gets overwritten with the host/port combination every time the emulator starts. It looks like env-init command reads from this file to generate "export PUBSUB_EMULATOR_HOST=localhost:8085" configuration.

So checking the file's existence is insufficient anyway -- the important thing is that the file *updated*. I'll try catching the output instead; this should work at least as well.

In principle, it should work on windows -- this is why I used Paths instead of hardcoding the directory.